packages/adblocker/src/engine/bucket/preprocessor.ts

```diff
@@ -0,0 +1,161 @@
+import { StaticDataView } from '../../data-view';
+import Preprocessor, { Env } from '../../preprocessor';
+
+export default class PreprocessorBucket {
+  public static deserialize(view: StaticDataView, env: Env): PreprocessorBucket {
+    const preprocessors: Preprocessor[] = [];
+
+    for (let i = 0, l = view.getUint32(); i < l; i++) {
+      preprocessors.push(Preprocessor.deserialize(view));
+    }
+
+    return new this({
+      env,
+      preprocessors,
+    });
+  }
+
+  public readonly conditions: Map<string, Preprocessor>;
+  public readonly ineligible: Set<number>;
+
+  public env: Env;
+
+  constructor({ env, preprocessors = [] }: { env: Env; preprocessors?: Preprocessor[] }) {
+    this.env = env;
+
+    this.ineligible = new Set();
+    this.conditions = new Map();
+
+    this.update({ added: preprocessors });
+  }
+
+  private build() {
```

Comment:
Could you explain why we need to keep mappings, etc.? Instead of something more straightforward like:
```js
const ineligible = new Set():
for (const preprocessor of this.preprocessors) {
  if (!preprocessor.evaluate(env)) {
    for (const filterId of preprocessor.filterIds) {
      ineligible.add(filterId);
    }
  }
}
```

Reply:
This is to simplify the evaluation of each filter. We need to collect preprocessors assigned to the filter and the constructed mapping does.